### PR TITLE
feature(macro): add a sub global toc with links

### PR DESCRIPTION
**Pull request by rgroux**
Hi,

like i said in #163, and for @calid, i have "Sub Global TOC" in y config.ru.

I made this PR and migrate the macro into macro lib folder in the gollum-lib, there are also an embryonic test (a copy of GlobalTOC).

### CODE REVIEW

**Comment by rgroux** [comment]
A short description of the PR:

This macro extend the GlobalTOC, with a second argument, if empty SubGlobalTOC and GlobalTOC are identical, but with  second argument, we can filter with match the page we want to expose in the TOC. it can be useful for TOC of subfolder or a TOC type of page.

SubGlobalTOC("(Pages in this Wiki", "") == SubGlobalTOC("(Pages in this Wiki", "") == GlobalTOC("(Pages in this Wiki")

SubGlobalTOC("(Pages in this Wiki", "^FirstLevel") => TOC from sub folder FirstLevel
SubGlobalTOC("(Pages in this Wiki", "Test") => all pages with Test in the name

hope it will be accepted, and used and hack/extend in a near future

**Comment by dometto** [comment]
@calid in #163:
> Maybe make the default root-at-current-page, and just override that if a parameter is passed? I think this would be really nice (and intuitive) default behavior

I agree with that. Would it be too much work to implement this in this PR?

**Comment by dometto** [comment]
Also I think the name `SubGlobalTOC` is a bit complicated and not very clear (since it's precisely not a global TOC). Maybe take up a suggestion by @flozzone and call it something like `Navigation()`?

**Comment by rgroux** [comment]
@dometto 
:+1:  for Navigation(), great name

for root-at-current-page, i will do it ASAP, today or tomorow.

**Comment by rgroux** [comment]
@dometto 

i think it's ok now

File: lib/gollum-lib/macro/navigation.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Gollum
+  class Macro
+    class Navigation < Gollum::Macro
+      def render(title = "Navigate in the TOC", toc_root_path = ::File.dirname(@page.path), full_path = false)
+        if @wiki.pages.size > 0
+          result = '<ul>' + @wiki.pages.map {|p|
```
Comment (bartkamphorst) [inline lib/gollum-lib/macro/navigation.rb]: As a stylistic suggestion, you might consider the following.
```ruby
list_items = @wiki.pages.map do |page|
  if page.url_path.start_with?(toc_root_path)
    path_display = full_path ? page.url_path_display  : page.url_path_display.sub(toc_root_path,"").sub(/^\//,'')
    "<li><a href=\"/#{page.url_path}\">#{path_display}</a></li>"
  end
result = "<ul>#{list_items.join}</ul>"
```

**Comment by rgroux** [comment]
@bartkamphorst 

yes, easy and more readable :)

**Comment by dometto** [comment]
:+1: Thanks!

**Comment by calid** [comment]
hmm so I just installed gollum-lib off of master, and unfortunately this is not working for me (TOC is always empty).  

It seems to be because `toc_root_path` always resolves to `.` but `page.url_path` resolves to the name of the page without a `./` prefix, e.g. `Home`.  

So this condition never passes and pages never get added to the TOC:

```ruby
if page.url_path.start_with?(toc_root_path)
    path_display = full_path ? page.url_path_display  : page.url_path_display.sub(toc_root_path,"").sub(/^\//,'')
    "<li><a href=\"/#{page.url_path}\">#{path_display}</a></li>"
end
```

To debug I added this else condition:
```ruby
else
    "<li>'#{toc_root_path}' != '#{page.url_path}'</li>"
end
```

which in my test wiki prints:
*    '.' != 'Home'
*    '.' != 'gollum/Gollum'
*     '.' != 'lists/ToDo'

Would it be too simplistic to just prefix `page.url_path` with `./` ?

**Comment by calid** [comment]
Actually, yeah it would... looks like even in nested wiki dirs toc_root_path still just resolves to `.`

**Comment by dometto** [comment]
Meh, apologies for the uncareful merge. Should have done some manual testing. @rgroux, any ideas for a fix?

**Comment by calid** [comment]
I wonder if #163 was prematurely closed.  I've been using @flozzone's branch locally and it works well, just missing the 'root at current subtree by default' logic.

@rgroux @flozzone's implementation seems more robust and already works with respect to pretty printing the TOC links as a tree.  Instead of reimplementing that logic what about just adding the subtree rooting and metadata title features to his implementation?

**Comment by dometto** [comment]
That would be my preference, too.

**Comment by rgroux** [comment]
@calid, @dometto 

sorry, for the bug. no idea actually, and i will be in a rush next week, i wil ltry a to look at it. but no guaranties.

little ideas for toc_root_path =::File.dirname(@page.path)
  - the wiki_path in config.ru
  - the wiki_path in command line
  - or any path in the config.

hope it will be this and easy to check and fix it

**Comment by polyzen** [comment]
Bump

**Pull request merged**